src/vfio/core.c

```diff
@@ -71,8 +83,47 @@ static int vfio_init_container(struct vfio_container *vfio)
 		return -1;
 	}
 
+	vfio->config_iommu = vfio_container_configure_iommu;
+	vfio->get_dev_fd = vfio_container_get_dev_fd;
+	vfio->do_dma_map = vfio_container_do_map_dma;
+	vfio->do_dma_unmap = vfio_container_do_unmap_dma;
+
 	return 0;
+}
+
+static int vfio_init_iommufd(struct vfio_container *vfio)
+{
+	vfio->fd = open("/dev/iommu", O_RDWR);
+	if (vfio->fd < 0) {
+		log_error("Error opening /dev/iommu\n");
+		return -1;
+	}
+
+	vfio->config_iommu = vfio_iommufd_configure_iommu;
+	vfio->get_dev_fd = vfio_iommufd_get_dev_fd;
+	vfio->do_dma_map = vfio_iommufd_do_map_dma;
+	vfio->do_dma_unmap = vfio_iommufd_do_unmap_dma;
+
+	return 0;
+}
 
+static int vfio_init_container(struct vfio_container *vfio)
+{
+	vfio->iommu_usrapi = getenv("VFN_USE_IOMMUFD") != NULL;
```

Comment:
We should probably turn this into build type discovery. That is, meson should check the availability of `linux/iommufd.h` and define something like `HAVE_LINUX_IOMMUFD_H`. We should then check on that with `#if`'s.

Reply:
yes